io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -98,6 +124,40 @@ selectTraceEvents' fn =
               []
     . traceSelectTraceEvents fn
 
+selectTraceRaces :: SimTrace a -> [ScheduleControl]
+selectTraceRaces = go
+  where
+    go (SimTrace _ _ _ _ trace)      = go trace
+    go (TraceRacesFound races trace) =
+      races ++ go trace
+    go _                             = []
+
+-- Extracting races from a trace.  There is a subtlety in doing so: we
+-- must return a defined list of races even in the case where the
+-- trace is infinite, and there are no races occurring in it! For
+-- example, if the system falls into a deterministic infinite loop,
+-- then there will be no races to find.
```

Comment:
I ran into this problem repeatedly testing the governor, and it was quite hard to debug. Limiting simulated time does not solve it, because some bugs (including the one I ran into) fall into a deterministic zero-time loop, so simulated time never advances. Limiting the number of trace events could work, if performed by the IOSim scheduler itself, but many tests are written to filter out a subset of trace events first--limiting the number of filtered events doesn't guarantee that you can avoid an infinite loop.
Also, race detection is quite expensive, and in practice we are only going to be able to explore a small subset of all possible races. Laziness is a really big win here: it allows us not even to compute the races that we have no time to explore--I originally had a strictness annotation on the races in the trace, and removing that strictness annotation dramatically sped up the simulations. So we really don't want a solution that forces us to compute all the races anyway.
I think we need the unsafePerformIO anyway, so we can compute just those races that affect the property we are testing.